### ISSUE TITLE
Make atomic peek()/poke() opt-in and mark them as unstable

### ISSUE THREAD

**Issue by ronawho** [chapel-lang/chapel#13827]
TL:DR we're not ready to stabilize `peek()`/`poke()` atomic operations, so I'm proposing we move them into an unstable opt-in module.

---

Atomic `peek()`/`poke()` are documented as `read()`/`write()` with relaxed memory ordering, but that's wrong. That's what we do for processor atomics, but for network atomics they map to a naked load/store of the underlying type.  This means that `peek()`/`poke()` are **not** atomic.

These operations are intended to limit the overhead of atomics (particularly network atomics) when atomicity is not required. For network atomics, even if an atomic variable lives on the current locale we normally perform operations through the NIC because processor and network atomics are not coherent. This means there is a large overhead for operations that do not require atomicity.  `peek()`/`poke()` were added to avoid this overhead when operations are known to be local and atomicity is not required. For example ra-atomics uses [`poke()`](https://github.com/chapel-lang/chapel/blob/0bb4bb460e4b4e20b420687ab8a5cae685a48e88/test/release/examples/benchmarks/hpcc/ra-atomics.chpl#L104) to initialize an atomic array and [`peek()`](https://github.com/chapel-lang/chapel/blob/0bb4bb460e4b4e20b420687ab8a5cae685a48e88/test/release/examples/benchmarks/hpcc/ra-atomics.chpl#L162) to perform a `+ reduce`.

There have been user requests asking for a way to drop the atomicity to improve performance during regions where atomicity is not required, and if we had a more general feature like that, we wouldn't need `peek()`/`poke()`.

I'm proposing that we move peek/poke into an opt-in `UnstableAtomics` module and deprecate the current available-by-default versions. I think we should also remove them from the spec. This allows us to defer the stabilization of peek/poke/non-atomic-ops, but still stabilize other atomic features.

A prototype implementation is available: https://github.com/chapel-lang/chapel/compare/master...ronawho:opt-in-peek-poke

(Suggestions for a better name than `UnstableAtomics` are welcome)

---

Aside about feature requests to support non-atomic operations on atomics:

There have been a few user requests to perform non-atomic updates on atomic variables for better performance when atomicity is not required. This seems attractive from a performance perspective, but I'm not sure how this would be implemented

In Rust you can get a mutable reference to the underlying type with [`get_mut()`](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicI8.html#method.get_mut) I believe Rust can achieve this because they only support atomics on platforms that have hardware support, so they can always implement atomics with native word sizes. See their note about [portability](https://doc.rust-lang.org/std/sync/atomic/index.html#portability)

This is not true for Chapel since we support atomics everywhere and have a portable locks backend. Personally I can't imagine Chapel without atomics support so it seems hard to only support them on platforms with hardware support.

So far as I know C/C++ do not have a way to drop the atomicity either, but this is not an area I'm very familiar with, so I could be wrong.

In addition to processor atomics whatever we do needs to work for network atomics too. All implementations that I know of (ugni, ofi, and gasnet-ex) all use native word sizes, but I'm not sure if that will be true in general

**Comment by ronawho** [comment]
As a quick performance comparison for why peek/poke (or the ability to drop atomicity) is attractive, here's a test that shows initializing an atomic array with `poke()` is ~200x faster than the using `write()` (under comm=ugni). There is less of an overhead for processor atomics, but it's still non-zero:

(This is just meant to show that the feature is attractive for performance, so we don't want to just get rid of it, but we need more time to decide on the final interface)


```chpl
use BlockDist, Time;

config const n = 20_000_000_000;
const D = {1..n} dmapped Block({1..n});

writef("| config                   | Time    |\n");
writef("| ------------------------ | ------- |\n");

{
  var netAtomicA: [D] atomic int;
  forall a in netAtomicA do a.poke(1);
  var t: Timer; t.start();

  forall a in netAtomicA do a.write(1);
  writef("| Network atomic write()   | %{###.##}s |\n", t.elapsed()); t.clear();

  forall a in netAtomicA do a.poke(1);
  writef("| Network atomic poke()    | %{###.##}s |\n", t.elapsed()); t.clear();
}

{
  var procAtomicA: [D] chpl__processorAtomicType(int);
  forall a in procAtomicA do a.poke(1);
  var t: Timer; t.start();

  forall a in procAtomicA do a.write(1);
  writef("| Processor atomic write() | %{###.##}s |\n", t.elapsed()); t.clear();

  forall a in procAtomicA do a.poke(1);
  writef("| Processor atomic poke()  | %{###.##}s |\n", t.elapsed()); t.clear();
}

{
  var nonAtomicA:  [D] int;
  forall a in nonAtomicA do a = 1;
  var t: Timer; t.start();

  forall a in nonAtomicA do a = 1;
  writef("| Non-atomic assign        | %{###.##}s |\n", t.elapsed()); t.clear();
}
```

```
chpl write-vs-poke.chpl --fast
./write-vs-poke -nl 4
```

| config                   | Time    |
| ------------------------ | ------- |
| Network atomic write()   | 372.26s |
| Network atomic poke()    |   1.83s |
| Processor atomic write() |   3.06s |
| Processor atomic poke()  |   1.84s |
| Non-atomic assign        |   0.73s |

**Comment by mppf** [comment]
I'd just put these in a Package module called PeekPoke.chpl.
